File: solution/038CountAndSay/038CountAndSay.go
package main

import "fmt"
import "bytes"

func countAndSayHelper(s string) string {
	var b bytes.Buffer
	i := 0
	j := 0
	c := byte(0)
	for j < len(s) {
		if s[i] == s[j] {
			c++
		} else {
			b.WriteByte('0' + c)
			b.WriteByte(s[i])
			i = j
			c = 1
		}
		j++
	}
	if c != 0 {
		b.WriteByte('0' + c)
		b.WriteByte(s[i])
	}
	return b.String()
}
func countAndSay(n int) string {
	if n <= 0 {
		return ""
	}
	s := "1"
	for i := 1; i < n; i++ {
		s = countAndSayHelper(s)
	}
	return s
}

func main() {
	fmt.Printf("%v\n", countAndSay(1))
	fmt.Printf("%v\n", countAndSay(2))
	fmt.Printf("%v\n", countAndSay(3))
	fmt.Printf("%v\n", countAndSay(4))
	fmt.Printf("%v\n", countAndSay(30))
}

File: solution/038CountAndSay/038CountAndSay.rs
impl Solution {
    fn count_and_say_helper(strs: String) -> String {
        let strs = strs.as_bytes();
        let n = strs.len();
        let mut i = 0;
        let mut j = 0;
        let mut count = 0;
        let mut buffer = String::new();
        while j < n {
            while j < n && strs[j] == strs[i] {
                j += 1;
                count += 1;
            }    
            buffer.push_str(count.to_string().as_str());
            buffer.push_str((strs[i] as char).to_string().as_str());
            count = 0;
            i = j;
        }
        buffer
    }
    
    pub fn count_and_say(n: i32) -> String {
        let mut buffer = "1".to_string();
        for i in 1..n {
            buffer = Self::count_and_say_helper(buffer);
        }
        buffer
    }
}
